modules/FvwmConsole/FvwmConsole.man
.\" t
.\" @(#)FvwmConsole.1	2/15/96
.TH FvwmConsole n "Feb 15 1996"
.UC
.SH NAME
FvwmConsole \- the FVWM command input interface
.SH SYNOPSIS
FvwmConsole is spawned by fvwm, so no command line invocation will work.
From within the .fvwm2rc file or from a menu, FvwmConsole takes any
options xterm will take, except -e, -name, and -title.

.RS
Module FvwmConsole -g 40x10 -fg black -bg green3
.RE

Also X resources can be set in .Xdefaults file

.RS
FvwmConsole*geometry: 40x4
.br
FvwmConsole*font: 7x14
.RE

.SH DESCRIPTION
The FvwmConsole allows the user to type fvwm commands into xterm, and
have them executed immediately. These commands are usually specfified
in the .fvwmrc file, or are bound to menu/mouse items as specified
in that file. This tools is particularly useful for testing new
configuration ideas, or for implementing temporary changes to your
environment.

GNU readline library is called to edit command line.
Also a perl script is supplied as alternative command editor. Its
commands are similar to emacs.

.RS
.PD 0
.TP 8
Cntl-A
- begining of line
.TP
Cntl-B
- previous char
.TP
Cntl-D
- delete char
.TP
Cntl-E
- end of line
.TP
Cntl-F
- next char
.TP
Cntl-H
- backspace
.TP
Cntl-K
- erase to the end of line
.TP
Cntl-N
- next line
.TP
Cntl-P
- previous line
.TP
Cntl-R
- search reverse
.TP
Cntl-U
- delete line
.TP
Meta-B
- previous word
.TP
Meta-F
- next word
.TP
Esc <
- begining of history
.TP
Esc >
- end of history
.RE
.PD

.SH COPYRIGHTS
Copyright 1996, Toshi Isogai. No guarantees or warranties are provided.
Use this program at your own risk. Permission to use this program for
any purpose is given, as long as the copyright is kept intact.


.SH INVOCATION
FvwmConsole can be invoked by inserting the line 'Module FvwmConsole' in
the .fvwmrc file. This can be placed on a line by itself, if FvwmConsole
is to be spawned during fvwm's initialization, or can be bound to a
menu or mouse button or keystroke to invoke it later. Fvwm will search
directory specified in the ModulePath configuration option to attempt
to locate FvwmConsole.

.SH CONFIGURATION OPTIONS
FvwmConsole uses xterm. All resources set for xterm are inherited
unless overiden by command line options.

.SH SEE ALSO
xterm(1), GNU readline library

.SH AUTHOR
Toshi Isogai  toshi_isogai@maxtor.com


ChangeLog
2026-02-23 David Flater

	Renamed package to Fvwm95 Forever and moved to GitHub.
	PACKAGE_VERSION = "Forever"

	Fixed FTBs for GCC 15.
	Updated things in configure.ac, Makefile.am, and system.fvwm95rc.in.
	Converted README to Markdown and revised as necessary.
	Merged BUGS and NEVERDO into README.md.

2020-12-10 David Flater

	fvwm95-2.0.43f-dwf8

	* Fixed multiple definitions of global variables in modules
	FvwmScroll and FvwmTaskBar.  GCC 10 changed the default behavior
	from -fcommon to -fno-common, which caused these multiple
	definitions to be rejected.

2020-03-18 David Flater

	fvwm95-2.0.43f-dwf7

	* Fixed buffer overflows in fvwm/read.c SendDataToModule which
	were triggered by a long IconPath or PixmapPath.

	* Added mini-firefox.xpm.

	* Updated system.fvwm95rc.in:
	- Added Firefox;
	- Added dosbox-x to the workaround for SDL2 fullscreen lossage;
	- Disabled Alt-arrow hotkeys (shortcuts to scroll by 1/10 of a
	page) to stop conflict with Doom (Alt = strafe).

	* Updated INSTALL.

2018-01-22 David Flater

	fvwm95-2.0.43f-dwf6

	* Updated system.fvwm95rc.in:
	- Fixed usage of killall;
	- Added the fullscreen workaround to chocolate-doom etc.;
	- Comments.

	* Updated INSTALL.

2017-04-20 David Flater

	fvwm95-2.0.43f-dwf5

	* Replaced faulty recursive build system with monolithic Automake:
        - Added script utils/makemakefileam.sh to generate Makefile.am;
	- Replaced configure.ac;
	- Purged all Imakefile and Makefile.in;
	- Adjusted install and search paths to agree with modern Linux.

	* Fixed innumerable instances of overflow, underflow, corruption,
	undefined behavior, division by zero, and broken parsing which
	were responsible for crashes and misbehavior in fvwm95 and every
	module.

	* Cleaned up compiler warnings through -Wall -Wno-parentheses on
	GCC 6.3.0.  Left the following:
	  ButtonArray.c:
	    warning: statement with no effect (BOGUS)
          etc:
  	    warning: the use of `mktemp' is dangerous, better use `mkstemp'
	    warning: implicit declaration of function 'ConsoleMessage' (BOGUS)

	* Harmonized swallow triggers in FvwmButtons, FvwmWharf,
	FvwmTaskBar, and FvwmScript (SwallowExec) to match on resource
	name (Resource), resource class (Class), or window name (Name),
	but not icon name (Icon Name).

	* Fixed fvwm95 bugs:
	- windows suddenly becoming enormous when resizing occurs near the
	  edge of the virtual desktop;
	- full-screen glitch when resizing a window that was declared with
  	  NoHandles and BorderWidth 0;
	- UseStyle option not accepting quoted identifiers;
	- false matching by UseStyle option;
	- no default WindowManagerName for Restart command (now fvwm95).

	* Fixed FvwmAudio bugs:
	- playing wrong or no sounds;
	- hanging, waiting on subprocesses to exit;
	- breaking audio_filename at whitespace.

	* Fixed FvwmConsole not using libreadline by default.

	* Fixed FvwmScript and sample scripts bugs:
	- missing default for FvwmScriptPath;
	- references to non-existent things in Buttons script;
	- infinite loop in PointerSetup script (but it is hopelessly fragile
	  and may hang on another system).

	* Fixed FvwmTaskBar bugs:
	- swallowing completely broken;
	- animation broken or too fast;
	- main event loop complexity;
	- no warning on missing *FvwmTaskBarGoodyLoadableSymbol;
	- bad geometry, location, and spacing of goodies;
	- bad calibration of x coordinates on mouse clicks on goodies;
	- area cleared different than area drawn;
	- missing default for FvwmTaskBarModuleIconPathBatStat;
	- only allowing one directory in FvwmTaskBarModuleIconPathBatStat.

	* Fixed FvwmWharf bugs:
	- black box over first button when swallowing failed;
	- wrong background colors on 32-bit arch;
	- animation too fast.

	* Corrected inaccuracies in man pages and HTML docs:
	- Documented the swallow commands for FvwmTaskBar;
	- Corrected description of the swallow commands for FvwmWharf;
	- Fixed ghost reference to FvwmInitBanner;
	- Corrected description of the name-matching behavior of the Style
  	  command to agree with what is actually done in fvwm/add_window.c
	  LookInList.

	* Made a few minor additions and enhancements:
	- Replaced the sample/default system.fvwm95rc;
	- Made FvwmTaskBar goodies work by single-click;
	- Improved error reporting by GoodyLoadable.c;
	- Added mini-xemacs.xpm, mini-chrome.xpm, mini-chromium.xpm,
	  mini-wilber16.xpm, and green.xpm;
	- Added LogoTest script.

	* Cleaned up random things:
	- Merged all needed preprocessor defines into FVWMconfig.h and made
	  all sources include it;
	- Removed version.h headers and declarations that conflicted with
	  FVWMconfig.h;
	- Merged index.html and GPL.html in with the other HTML files;
	- Merged xpmroot into utils;
	- Merged and purged duplicative icon collections:
	  - FvwmScript/Scripts/some_nice_icons:
	      moved shutdown.xpm to pixmaps/shutdown2.xpm
	      moved window.xpm to pixmaps/window2.xpm
	  - FvwmTaskBar/icons:
	      moved mini-mail.xpm to mini-icons/mini-mail2.xpm
	      moved minimail.xbm to mini-icons/minimail.xbm
	- Fixed saferealloc and made UpdateString use it (3 copies);
	- Moved random READMEs, example files, etc. from module
	  subdirectories into docs/modules;
	- Moved other random READMEs, example files, etc. into docs/random;
	- Moved the untested/unmaintained ASSound and OffiX subdirs of
	  FvwmWharf to docs/modules/FvwmWharf;
	- Deleted all other unused source files and scripts;
	- Normalized file permissions;
	- Stripped trailing whitespace.

	Not done:  see BUGS, NEVERDO.

2014-01-26 David Flater

	fvwm95-2.0.43f-dwf4

	* Fixed BatStatModule showing "charging" icon instead of "online"
	when fully charged.

	* Changed the "alert" and "empty" levels from 10% and 5% to 7% and
	2% respectively.

2014-01-25 David Flater

 	fvwm95-2.0.43f-dwf3

	* Fixed crash of FvwmTaskBar when a goody (BatStatModule) returned
	NULL from its init function to signify failure to init.

	* Updated BatStatModule to work with /sys/class/power_supply.

modules/FvwmConsole/FvwmConsole.c
#include <FVWMconfig.h>
/*
  Fvwm command input interface.

  Copyright 1996, Toshi Isogai. No guarantees or warantees or anything
  are provided. Use this program at your own risk. Permission to use
  this program for any purpose is given,
  as long as the copyright is kept intact.
*/

#include "FvwmConsole.h"

char *MyName;

int fd[2];  /* pipe to fvwm */
FILE *sp;
int  s,ns;             /* socket handles */
char name[32]; /* name of this program in executable format */
int  pid;      /* server routine child process id */

void server(int *fd);
void GetResponse();
void DeadPipe(int signum);
void CloseSocket();
void ErrMsg( char *msg );
void SigHandler(int signum);

#define XARGS (sizeof(xterm_a)/sizeof(char *))

int main(int argc, char **argv){
  char *tmp, *s;
  static char client[120];
  char **eargv;
  int i,j;
  char *xterm_a[] = {"-title", name, "-name", name, "-e", client, NULL};

  /* initially no child */
  pid = 0;

  /* Save the program name - its used for error messages and option parsing */
  tmp = argv[0];

  s=strrchr(argv[0], '/');
  if (s != NULL)
    tmp = s + 1;

  strcpy( name, tmp );

  MyName = safemalloc(strlen(tmp)+2);
  strcpy(MyName,"*");
  strcat(MyName, tmp);

  /* construct client's name */
  strcpy( client, argv[0] );
  strcat( client, "C" );

  if(argc < FARGS)    {
	fprintf(stderr,"%s Version %s should only be executed by fvwm!\n",
                       MyName, VERSION);
	exit(1);
  }

  // 2017-03-11 DWF Fixed short by one (0 is for xterm command).
  if( ( eargv =(char **)safemalloc((argc-FARGS+XARGS+1)*sizeof(char *)) ) == NULL ) {
	ErrMsg( "allocation" );
  }

  /* copy arguments */
  eargv[0] = XTERM;
  j= 1;
  for ( i=FARGS ; i<argc; i++,j++ ) {
	eargv[j] = argv[i];
  }

  for ( i=0 ; xterm_a[i] != NULL ; j++, i++ ) {
	eargv[j] = xterm_a[i];
  }
  eargv[j] = NULL;

  /* Dead pipes mean fvwm died */
  signal (SIGPIPE, DeadPipe);
  signal (SIGINT, SigHandler);
  signal (SIGQUIT, SigHandler);

  fd[0] = atoi(argv[1]);
  fd[1] = atoi(argv[2]);

  /* launch xterm with client */
  if( fork() == 0 ) {
	execvp( *eargv, eargv );
	ErrMsg("exec");
  }

  server(fd);

  return 0;
}

/***********************************************************************
 *	signal handler
 ***********************************************************************/
void DeadPipe(int signum) {
  fprintf(stderr,"%s: dead pipe\n", name);
  CloseSocket();
  exit(0);
}

void SigHandler(int signum) {
  CloseSocket();
  exit(1);
}

/*********************************************************/
/* close sockets and spawned process                     */
/*********************************************************/
void CloseSocket() {
  if( pid ) {
	kill( pid, SIGKILL );
  }
  close(ns);     /* remove the socket */
  fclose(sp);
  unlink( S_NAME );
}

/*********************************************************/
/* setup server and communicate with fvwm and the client */
/*********************************************************/
void server (int *fd) {
  struct sockaddr_un sas, csas;
  int  len;     /* length of sockaddr */
  socklen_t clen;
  char buf[BUFSIZE];      /*  command line buffer */

  /* make a socket  */
  if( (s = socket(AF_UNIX, SOCK_STREAM, 0 )) < 0  ) {
	ErrMsg( "socket");
	exit(1);
  }

  /* name the socket */
  sas.sun_family = AF_UNIX;
  strcpy( sas.sun_path, S_NAME );

  /* bind the above name to the socket */
  /* first, erase the old socket */
  unlink( S_NAME );
  len = sizeof( sas.sun_family) + strlen( sas.sun_path );

  if( bind(s, (struct sockaddr *)&sas, len) < 0 ) {
	ErrMsg( "bind" );
	exit(1);
  }

  /* listen to the socket */
  /* set backlog to 5 */
  if ( listen(s,5) < 0 ) {
    ErrMsg( "listen" );
	exit(1);
  }

  /* accept connections */
  clen = sizeof(csas);
  if(( ns = accept(s, (struct sockaddr *)&csas, &clen)) < 0 ) {
	ErrMsg( "accept");
	exit(1);
  }

  /* get command from client and return result */
  sp = fdopen( ns, "r" );
  pid = fork();
  if( pid == -1 ) {
	ErrMsg(  "fork");
	exit(1);
  }
  if( pid == 0 ) {
	while(fgets( buf, BUFSIZE, sp )) {

	  /* check if client is terminated */
	  if( buf == NULL ) {
		break;
	  }
	  SendText(fd,buf,0); /* send command */
	}
	CloseSocket();
	kill( getppid(), SIGQUIT);
	exit(0);
  }
  while(1) {
	GetResponse();
  }
}

/**********************************************/
/* read fvwm packet and pass it to the client */
/**********************************************/
void GetResponse() {
  fd_set in_fdset;
  unsigned long *body;
  unsigned long header[HEADER_SIZE];

  FD_ZERO(&in_fdset);
  FD_SET(fd[1],&in_fdset);

  /* ignore anything but error message */
  if( ReadFvwmPacket(fd[1],header,&body) > 0)	 {
	if(header[1] == M_PASS)	     {
	  send( ns, (char *)&body[3], strlen((char *)&body[3]), 0 );
	}
	free(body);
  }
}

/************************************/
/* print error message on stderr */
/************************************/
void ErrMsg( char *msg ) {
  fprintf( stderr, "%s server error in %s\n", name, msg );
  CloseSocket();
  exit(1);
}
